src/memory-analyzer/analyze_symbol.cpp

```diff
@@ -0,0 +1,375 @@
+  PRECONDITION(expr.type() == zero_expr.type());
+
+  std::string c_expr = c_converter.convert(expr);
+  const pointer_valuet value = gdb_api.get_memory(c_expr);
+
+  const auto memory_location = value.address;
+
+  if(!memory_location.is_null())
+  {
+    if(is_c_char_type(expr.type().subtype()))
+    {
+      return get_char_pointer_value(expr, memory_location, location);
+    }
+    else
+    {
+      const exprt target_expr =
+        get_non_char_pointer_value(expr, memory_location, location);
+
+      if(target_expr.id() == ID_nil)
+      {
+        outstanding_assignments[expr] = memory_location;
+      }
+      else
+      {
+        return address_of_exprt(target_expr);
+      }
+    }
+  }
+
+  return zero_expr;
+}
+
+exprt gdb_value_extractort::get_array_value(
+  const exprt &expr,
+  const exprt &array,
+  const source_locationt &location)
+{
+  PRECONDITION(array.id() == ID_array);
+
+  PRECONDITION(expr.type().id() == ID_array);
+  PRECONDITION(expr.type() == array.type());
+
+  exprt new_array(array);
+
+  for(size_t i = 0; i < new_array.operands().size(); ++i)
+  {
+    const index_exprt index_expr(expr, from_integer(i, index_type()));
+
+    exprt &operand = new_array.operands()[i];
+
+    operand = get_expr_value(index_expr, operand, location);
+  }
+
+  return new_array;
+}
+
+exprt gdb_value_extractort::get_expr_value(
+  const exprt &expr,
+  const exprt &zero_expr,
+  const source_locationt &location)
+{
+  PRECONDITION(expr.type() == zero_expr.type());
+
+  const typet &type = expr.type();
+
+  if(is_c_integral_type(type))
+  {
+    INVARIANT(zero_expr.is_constant(), "zero initializer is a constant");
+
+    return from_integer(string2integer(get_gdb_value(expr)), expr.type());
+  }
+  else if(is_c_char_type(type))
+  {
+    INVARIANT(zero_expr.is_constant(), "zero initializer is a constant");
+
+    return zero_expr; // currently left at 0
+  }
+  else if(type.id() == ID_c_bool)
+  {
+    INVARIANT(zero_expr.is_constant(), "zero initializer is a constant");
+
+    return from_c_boolean_value(id2boolean(get_gdb_value(expr)), type);
+  }
+  else if(type.id() == ID_c_enum)
+  {
+    INVARIANT(zero_expr.is_constant(), "zero initializer is a constant");
+
+    return convert_member_name_to_enum_value(
+      get_gdb_value(expr), to_c_enum_type(type));
+  }
+  else if(type.id() == ID_struct || type.id() == ID_struct_tag)
+  {
+    return get_struct_value(expr, zero_expr, location);
+  }
+  else if(type.id() == ID_array)
+  {
+    return get_array_value(expr, zero_expr, location);
+  }
+  else if(type.id() == ID_pointer)
+  {
+    INVARIANT(zero_expr.is_constant(), "zero initializer is a constant");
+
+    return get_pointer_value(expr, zero_expr, location);
+  }
+  UNREACHABLE;
+}
+
+exprt gdb_value_extractort::get_struct_value(
+  const exprt &expr,
+  const exprt &zero_expr,
+  const source_locationt &location)
+{
+  PRECONDITION(zero_expr.id() == ID_struct);
+
+  PRECONDITION(expr.type().id() == ID_struct_tag);
```

Comment:
Hmm, this can also be called with `expr.type().id() == ID_struct`, although I do think that that should actually never be the case, i.e., the `ID_struct` case should be removed from the call site.

Reply:
I added a precondition to `get_expr_value` and removed the case from its analysis.